python/tank/bootstrap/manager.py

```diff
@@ -626,35 +626,68 @@ def _is_toolkit_activated_in_shotgun(self):
         log.debug("...enabled: %s" % enabled)
         return enabled
 
-    def _cache_apps(self, tk, progress_callback, do_post_install=False):
+    def _cache_apps(self, tk, config_engine_name, config_project_id, progress_callback, do_post_install=False):
         """
         Caches all apps associated with the given toolkit instance.
 
         :param tk: Bootstrapped :class:`~sgtk.Sgtk` instance to cache items for.
+        :param config_engine_name: Name of the engine that was used to resolve the configuration.
+        :param config_project_id: Project id that was used to resolve the configuration;
+                                  None for the site configuration.
         :param progress_callback: Callback function that reports back on the engine startup progress.
         :param do_post_install: Set to true to execute the post install triggers.
         """
 
+        from ..platform import constants as platform_constants
+
         log.info("Downloading and installing apps...")
 
+        # Resolve a context for the project.
+        if config_project_id:
+            project_entity = {"type": "Project", "id": config_project_id}
+            project_context = tk.context_from_entity_dictionary(project_entity)
+        else:
+            project_context = tk.context_empty()
+
         # each entry in the config template contains instructions about which version of the app
         # to use. First loop over all environments and gather all descriptors we should download,
         # then go ahead and download and post-install them
         pc = tk.pipeline_configuration
 
+        try:
+            # Get an environment name given the project context.
+            env_name = tk.execute_core_hook(platform_constants.PICK_ENVIRONMENT_CORE_HOOK_NAME,
+                                            context=project_context)
+        except Exception, e:
+            log.debug("The pick environment core hook for context '%s' reported error: %s" % (project_context, e))
```

Comment:
It would be great if we could add something like 'Could not evaluate which environment to start up. Will download all dependencies for the entire config.'